cmd/iam-k8s-store.go

```diff
@@ -0,0 +1,554 @@
+	var basePath string
+	if isGroup {
+		basePath = iamConfigPolicyDBGroupsPrefix
+	} else {
+		switch userType {
+		case srvAccUser:
+			basePath = iamConfigPolicyDBServiceAccountsPrefix
+		case stsUser:
+			basePath = iamConfigPolicyDBSTSUsersPrefix
+		default:
+			basePath = iamConfigPolicyDBUsersPrefix
+		}
+	}
+	result := iamK8s.listIAMConfigs(ctx, basePath, false)
+	if result.err != nil {
+		return result.err
+	}
+	for _, item := range result.iamConfigItems {
+		policyFile := item.objPath
+		var p MappedPolicy
+		if err := json.Unmarshal([]byte(item.data), &p); err != nil {
+			return err
+		}
+		m[policyFile] = p
+	}
+	return nil
+}
+
+// Refresh IAMSys. If an object layer is passed in use that, otherwise load from global.
+func (iamK8s *IAMK8sStore) loadAll(ctx context.Context, sys *IAMSys) error {
+	return sys.Load(ctx, iamK8s)
+}
+
+func (iamK8s *IAMK8sStore) savePolicyDoc(ctx context.Context, policyName string, p iampolicy.Policy) error {
+	return iamK8s.saveIAMConfig(ctx, &p, getPolicyDocPath(policyName))
+}
+
+func (iamK8s *IAMK8sStore) saveMappedPolicy(ctx context.Context, name string, userType IAMUserType, isGroup bool, mp MappedPolicy, opts ...options) error {
+	return iamK8s.saveIAMConfig(ctx, mp, getMappedPolicyPath(name, userType, isGroup), opts...)
+}
+
+func (iamK8s *IAMK8sStore) saveUserIdentity(ctx context.Context, name string, userType IAMUserType, u UserIdentity, opts ...options) error {
+	return iamK8s.saveIAMConfig(ctx, u, getUserIdentityPath(name, userType), opts...)
+}
+
+func (iamK8s *IAMK8sStore) saveGroupInfo(ctx context.Context, name string, gi GroupInfo) error {
+	return iamK8s.saveIAMConfig(ctx, gi, getGroupInfoPath(name))
+}
+
+func (iamK8s *IAMK8sStore) deletePolicyDoc(ctx context.Context, name string) error {
+	err := iamK8s.deleteIAMConfig(ctx, getPolicyDocPath(name))
+	if err == errConfigNotFound {
+		err = errNoSuchPolicy
+	}
+	return err
+}
+
+func (iamK8s *IAMK8sStore) deleteMappedPolicy(ctx context.Context, name string, userType IAMUserType, isGroup bool) error {
+	err := iamK8s.deleteIAMConfig(ctx, getMappedPolicyPath(name, userType, isGroup))
+	if err == errConfigNotFound {
+		err = errNoSuchPolicy
+	}
+	return err
+}
+
+func (iamK8s *IAMK8sStore) deleteUserIdentity(ctx context.Context, name string, userType IAMUserType) error {
+	err := iamK8s.deleteIAMConfig(ctx, getUserIdentityPath(name, userType))
+	if err == errConfigNotFound {
+		err = errNoSuchUser
+	}
+	return err
+}
+
+func (iamK8s *IAMK8sStore) deleteGroupInfo(ctx context.Context, name string) error {
+	err := iamK8s.deleteIAMConfig(ctx, getGroupInfoPath(name))
+	if err == errConfigNotFound {
+		err = errNoSuchGroup
+	}
+	return err
+}
+
+func filterExpiredItems(items []iamConfigItem) []iamConfigItem {
+	keep := []string{}
+	for _, item := range items {
+		if hasTtlPrefix(item.objPath) {
+			expTime, _ := strconv.ParseInt(item.data, 10, 64)
+			if expTime > time.Now().Unix() {
```

Comment:
Potential race where some entries might be treated as expired when they have the same timestamp as others. I would move `time.Now().Unix()` to a variable before the for loop starts.

Is it likely to happen? no. I experienced something similar earlier, but I'm uncertain if this matters for your case.